### PR TITLE
Merge Geth v1.13.15

### LINKED ISSUE #188: Update Geth to v1.13+
[v1.14.0](https://github.com/ethereum/go-ethereum/releases/tag/v1.14.0) is already there for three days, but it's not supposed to be used in pre-merge (although need to reensure). Thus, we need to update to v1.13+.

This change may be incompatible for T3, need to check it.

### LINKED ISSUE #291: Mismatching block hash and transaction receipt data
#### System information

Geth version: `v0.2.1`
OS & Version: Windows/Linux/OSX

#### Expected behaviour

The `blockHash` field of a transaction receipt should be the hash of the block that contains this transaction.

When we fetch the receipt of transaction `0xdad9de0271d263e061c031a4bf5ac58355e16149e035bbfa5bb8e97a6f9449a1` from mainnet through RPC method `eth_getTransactionReceipt`, we will get the following response.

```json
{
    "blockHash": "0xfea21c66977db7b4fb33283e0e442c67321cd690f33da5f880fe02e16df6788e",
    "blockNumber": "0x11045",
    "contractAddress": "0x9ddb3368751af25e7af9a8c3ed92ea5dd5cfbe4a",
    "cumulativeGasUsed": "0xf023d",
    "effectiveGasPrice": "0x9502f9000",
    "from": "0xb5b184b4a5304be50e45f9489df9bce1cfc40f90",
    "gasUsed": "0xf023d",
    "logs": [],
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "status": "0x1",
    "to": null,
    "transactionHash": "0xdad9de0271d263e061c031a4bf5ac58355e16149e035bbfa5bb8e97a6f9449a1",
    "transactionIndex": "0x0",
    "type": "0x2"
}
```

#### Actual behaviour

However, the block `0xfea21c66977db7b4fb33283e0e442c67321cd690f33da5f880fe02e16df6788e` was reorged at the height of `69701`, and the final hash of that block is `0x1bb7685c5d85c818458858ad690c07b41a73a58bfa7e81e9182c64955ec05bf5`.

Because when we fetch the hash of block `0x11045` from mainnet through `eth_getBlockByNumber`, we will get the following response.

```json
{
    "baseFeePerGas": "0x4a817c800",
    "difficulty": "0x2",
    "extraData": "0x0034a3b2abb99b4c128acf61dcbbd1fcac0b1616524fe8af0dbb633283d8e9703668142fd130f2818d641ec1c538fa17e6ad8193c9b580f6850b114280763452f65353fffe73d46539e51a6ddfc0e2c86aa61ac4a4f006f4fceeb72ee0012a2d3367168d10e3973f57e8a0aa312c1917ab0e6a05d8b6af6609e6d1a9db6a0893926bd81c0ef93aaaa543c116f09bd475aca0317e8181a43c4f9e430708cebc4a93e633aa6c599e82061538e979658444efbe96abadff286914f47721a367c48bbcefc295d5d113e26e523d8c4900629a4336162ada25fc36fbd0f796e2377ccc1a7692d15a4b1511b2d365748df7538275f0bf0a61b44717745fa2c8babb9b696dad0cb92fcd5b99c20b9dc3a39f0063995c814cbdf9615a38a6f9d4de210c16bfff7629df4b15ed468e0d918e039362d2a8e036ab150ecf1a9268b89e2f6b848e4da2bfa551d85f18fdee1213a23300bbbe6e3972099089dbab41269303d98779a327c509fdc9c4fc95ebbe67c5d882421d5f17b6865daf34ad346f62e97d117f32639eae98071f2f0b5a9bc2b8581f0059ca291535dbc673f13a222d3814f41735eca5ff3309958b1f4ecae3402e5c8c17ea6bb155b00c69174d96e9d2ce689aaeded4d3341143d13877d1158803a29101",
    "gasLimit": "0x1c9c380",
    "gasUsed": "0xf023d",
    "hash": "0x1bb7685c5d85c818458858ad690c07b41a73a58bfa7e81e9182c64955ec05bf5",
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
    "miner": "0x1212000000000000000000000000000000000003",
    "mixHash": "0xeb59c093e3a02bfa4e0d4677d4769022cd9399bbc8b93ad1e892acd6a08aa533",
    "nonce": "0x0000000000000002",
    "number": "0x11045",
    "parentHash": "0x11f6a478dba0d59c85736264929c35dfd34cb5bd00976fff239971dffe161a76",
    "receiptsRoot": "0x89058005597e2b4b5e888c3efb707090f0944e1d66b4f8ab941b934075a3f178",
    "sha3Uncles": "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347",
    "size": "0x11b1",
    "stateRoot": "0x067d20bb2eef5818da7b4c75043630d3bbded3943e3100a88f931addd2f8df74",
    "timestamp": "0x66ae1404",
    "totalDifficulty": "0x21f53",
    "transactions": [
        "0xdad9de0271d263e061c031a4bf5ac58355e16149e035bbfa5bb8e97a6f9449a1"
    ],
    "transactionsRoot": "0xb353d8b4b112bd85f08597be61a9534ce6928bf40f842a7b6b6f1f3ca6ad358d",
    "uncles": [],
    "withdrawals": [],
    "withdrawalsRoot": "0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421"
}
```

#### Steps to reproduce the behaviour

Send above RPC requests to any of our running nodes.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
This PR merges [v1.13.15](https://github.com/ethereum/go-ethereum/releases/tag/v1.13.15) of Geth into bane-main and aimed to close #188.

Close #291.

- [X]  The node functions normally and consensus runs fine (1/4/7 privnet).
- [X] Mainnet/T4 sync is OK.

UPD on the compatibility problem described earlier: there's actually no compatibility problems with the 1.13.15 version merge. Turns out that Geth v1.13.15 was an unclean release from non-master branch. It's based on v1.13.14 with five additional commits cherry-picked (ref. https://github.com/ethereum/go-ethereum/pull/29556, the release branch itself is https://github.com/ethereum/go-ethereum/tree/release/1.13).

The list of important PRs that were merged. @txhsl, @chenquanyu, please take a look at these PRs before the review.
* Fix the enforced minimum miner tip (https://github.com/ethereum/go-ethereum/pull/28933).
* Add Cancun timestamps for mainnet (https://github.com/ethereum/go-ethereum/pull/28958). This PR may be useful for us while configuring our own schedule for Cancun.
* Genesis allocations were moved to a separate package (https://github.com/ethereum/go-ethereum/pull/29003). This PR does not affect us directly, but we need to be aware of this change.
* **An ability to pre-filter mempooled transactions by dynamic fee is added** (https://github.com/ethereum/go-ethereum/pull/29005). This change affects the #166, it needs to be adjusted accordingly.
* **Override block's BaseFee in eth_call API** (https://github.com/ethereum/go-ethereum/pull/29051). This change seems not to affect us directly, but it would be nice if you double-check it.

Also, it would be nice if you double-check the merged commits, maybe you'll find something important that we should react to.

### CODE REVIEW

**Comment by AnnaShaleva** [comment]
Tests pass, but privnet isn't running:
```
INFO [08-09|19:38:45.069] HTTP server started                      endpoint=[::]:8562 auth=false prefix= cors= vhosts=*
INFO [08-09|19:38:45.070] WebSocket enabled                        url=ws://[::]:8572
INFO [08-09|19:38:45.071] WebSocket enabled                        url=ws://127.0.0.1:8552
INFO [08-09|19:38:45.071] HTTP server started                      endpoint=127.0.0.1:8552 auth=true  prefix= cors=localhost vhosts=localhost
INFO [08-09|19:38:46.638] Unlocked account                         address=0x74f4EFFb0B538BAec703346b03B6d9292f53A4CD
INFO [08-09|19:38:46.638] Legacy pool tip threshold updated        tip=0
INFO [08-09|19:38:46.638] Legacy pool tip threshold updated        tip=1,000,000,000
INFO [08-09|19:38:46.638] Initializing BFT consensus               account=0x74f4EFFb0B538BAec703346b03B6d9292f53A4CD
INFO [08-09|19:38:46.638] Snap sync complete, auto disabling
INFO [08-09|19:38:46.638] Fetching latest sealing proposal         "desired number"=1
INFO [08-09|19:38:46.639] Commit new sealing work                  number=1 sealhash=e66155..adaa87 "parent hash"=edba97..2eac00 etherbase=0x1212000000000000000000000000000000000003 txs=0 gas=0 fees=0 elapsed="171.163µs"
INFO [08-09|19:38:47.640] Sealing proposal updated                 number=1 sealhash=e66155..adaa87 "parent hash"=edba97..2eac00 txs=0
INFO [08-09|19:38:47.640] Starting dBFT engine                     "last height"=0 "last timestamp"=0
panic: failed to retrieve next block validators: failed to perform 'getCurrentConsensus' call: execution reverted

goroutine 4252 [running]:
github.com/ethereum/go-ethereum/consensus/dbft.New.func4({0x0, 0xc75aa85fa0766dab?, 0x1304c9266330c603?})
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:326 +0x216
github.com/nspcc-dev/dbft.(*Context[...]).reset(0x1f8e0a0, 0x0, 0x2c)
	github.com/nspcc-dev/dbft@v0.2.0/context.go:201 +0xba
github.com/nspcc-dev/dbft.(*DBFT[...]).initializeConsensus(0x1f92ba0, 0x0, 0xc00012afc8)
	github.com/nspcc-dev/dbft@v0.2.0/dbft.go:88 +0x45
github.com/nspcc-dev/dbft.(*DBFT[...]).Start(0x1f92ba0, 0x0)
	github.com/nspcc-dev/dbft@v0.2.0/dbft.go:74 +0x67
github.com/ethereum/go-ethereum/consensus/dbft.(*DBFT).Start(0x0?, {0x1f78150?, 0xc000427808?})
	github.com/ethereum/go-ethereum/consensus/dbft/dbft.go:1133 +0x3eb
created by github.com/ethereum/go-ethereum/eth.(*Ethereum).StartMining in goroutine 1
	github.com/ethereum/go-ethereum/eth/backend.go:524 +0x6c9
```


I'm fixing it. @txhsl, @chenquanyu from your side regarding base fee changes I need you to take a look at https://github.com/ethereum/go-ethereum/pull/28933 and https://github.com/ethereum/go-ethereum/pull/29005 and check whether we need to adjust fees related code. From my PoW we can go without any additional changes, but may be I'm wrong.

**Comment by txhsl** [comment]
> @txhsl, @chenquanyu from your side regarding base fee changes I need you to take a look at https://github.com/ethereum/go-ethereum/pull/28933 and https://github.com/ethereum/go-ethereum/pull/29005 and check whether we need to adjust fees related code.

Yes, I think we can go without any additional changes. I'm glad to see that a minimum fee setting was finally applied to the blob pool, it made the codes reasonable.

But before we enable the blob transactions when release v0.2.2, we need to apply our policy limit to the pool properly, ref #266.

**Comment by AnnaShaleva** [comment]
> panic: failed to retrieve next block validators: failed to perform 'getCurrentConsensus' call: execution reverted

Regarding this problem:
1. Mainnet/testnet synchronisation fails with the same error on the top of the current version of PR.
2. I tried to locate the source of this error, and it seems to me that it goes from the Governance contract code itself. Given the fact that privnet is running OK on both current master (eef0b9643f65ca5b508da7b5c40051c29d8d7785) and on old branch with v1.13.15 changes on the top of it (e828bfcac3ab73a138b1939de5ea8f047824ff96) I suspect that something has been changed in both Geth v1.13.11+ and compiled Governance version which leads to the execution revert.
3. I located the commit starting from which privnet fails with this error. It's ee1cfc8, in this commit Governance contract (and a couple of other contracts) were updated to fetch recent changes.
4. Now I'm trying to locate commit from Geth v1.13.11+ that leads to this error and connected with changes in Governance introduced by ee1cfc8. I suppose that it should be some change in the VM opcode processing, may be yet another bug fix, or something related to Shanghai opcodes processing, because ee1cfc8 includes a compiler target switch (to Shanghai).

**Comment by txhsl** [comment]
> panic: failed to retrieve next block validators: failed to perform 'getCurrentConsensus' call: execution reverted

It is hard to figure out the reason, since the only log we can see is `execution reverted`. I'll check this contract calling from Cancun version as well.

**Comment by AnnaShaleva** [comment]
> I'll check this contract calling from Shanghai version as well.

Just to clarify: I don’t think it’s the contract’s fault. It looks like something has been changed in the VM processors.

**Comment by txhsl** [comment]
Just some results, I tried several test cases about `eth_Call` based on both `bane-main` and this branch. I sent `getCurrentConsensus()` request to different addresses, so the method hash is `0x9f9d7f81`.

||Address without any code|Contract with the method|Proxy contract without the method|Normal contract without the method|
|--|--|--|--|--|
|`bane-main`|`0x`|`{correct data}`|`execution reverted`|`execution reverted`|
|`upd-v1.13.15`|`0x`|`execution reverted`|`execution reverted`|`invalid opcode: PUSH0`|

File: params/config.go [hunk 1]
```diff
@@ -970,9 +973,9 @@ func (c *ChainConfig) Rules(num *big.Int, isMerge bool, timestamp uint64) Rules
 		IsBerlin:         c.IsBerlin(num),
 		IsLondon:         c.IsLondon(num),
 		IsMerge:          isMerge,
-		IsShanghai:       c.IsShanghai(num, timestamp),
-		IsCancun:         c.IsCancun(num, timestamp),
-		IsPrague:         c.IsPrague(num, timestamp),
-		IsVerkle:         c.IsVerkle(num, timestamp),
+		IsShanghai:       isMerge && c.IsShanghai(num, timestamp),
```
Comment (txhsl) [inline params/config.go]: Delete this `isMerge &&` and things will work.

Comment (AnnaShaleva) [reply params/config.go]: Oh gosh...You're right, it works.

**Comment by AnnaShaleva** [comment]
Let me do the final checks with mainnet/testnet sync.

**Comment by AnnaShaleva** [comment]
Mainnet sync passes:
```
2024-08-14T19:27:06.130+0300	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 2, "height": 144802, "view": 0, "my_height": 144802, "my_view": 0}
2024-08-14T19:27:06.130+0300	INFO	dbft@v0.2.0/dbft.go:498	received Commit	{"validator": 2}
2024-08-14T19:27:06.131+0300	INFO	dbft@v0.2.0/check.go:65	approving block	{"height": 144802, "hash": "0xf98bcbefc6f8b9972d904dbe18e0ddba60d5750c166b29420f8c92e469bd070e", "tx_count": 0, "merkle": "0x6c42725a23c7319177fbee1a7b737684ea07a1268e565766efd30983d9d62082", "prev": "0x86cdb96a0f2139db487f326f4c9cb831b0b65a984ac18d8909e32921ba04bc91"}
INFO [08-14|19:27:06.141] Chain reorg detected                     number=144,800 hash=2363f2..b0269d drop=1 dropfrom=5fea5a..bdd62e add=2 addfrom=9b648e..79888f
INFO [08-14|19:27:06.143] Successfully wrote new block with state  number=144,802 hash=9b648e..79888f
INFO [08-14|19:27:06.144] New block in the chain                   "dbft index"=144,802 "chain index"=144,802 hash=0x9b648e17fb85f30d3883fdabe6ed39568182b2f44e4c5377cad2406cd879888f "parent hash"=0x86cdb96a0f2139db487f326f4c9cb831b0b65a984ac18d8909e32921ba04bc91 primary=0 coinbase=0x1212000000000000000000000000000000000003 "mix digest"=0xeb59c093e3a02bfa4e0d4677d4769022cd9399bbc8b93ad1e892acd6a08aa533
INFO [08-14|19:27:06.144] Fetching latest sealing proposal         "desired number"=144,803
INFO [08-14|19:27:06.145] Commit new sealing work                  number=144,803 sealhash=416eab..b0ee09 "parent hash"=9b648e..79888f etherbase=0x1212000000000000000000000000000000000003 txs=0   gas=0 fees=0 elapsed=1.316ms
```
Testnet sync passes, although testnet sync is significantly slower than it should be due to #296, will fix it in another PR.
```
2024-08-14T20:04:37.689+0300	DEBUG	dbft@v0.2.0/check.go:56	not enough to commit	{"count": 3}
2024-08-14T20:04:37.704+0300	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 0, "height": 262743, "view": 0, "my_height": 262743, "my_view": 0}
2024-08-14T20:04:37.705+0300	INFO	dbft@v0.2.0/dbft.go:498	received Commit	{"validator": 0}
2024-08-14T20:04:37.705+0300	DEBUG	dbft@v0.2.0/check.go:56	not enough to commit	{"count": 4}
2024-08-14T20:04:37.709+0300	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 3, "height": 262743, "view": 0, "my_height": 262743, "my_view": 0}
2024-08-14T20:04:37.710+0300	INFO	dbft@v0.2.0/dbft.go:498	received Commit	{"validator": 3}
2024-08-14T20:04:37.710+0300	INFO	dbft@v0.2.0/check.go:65	approving block	{"height": 262743, "hash": "0xc57abd59633ab12a21ba5cbe948d43c461b2bc014687588b238cb47e87448ae8", "tx_count": 0, "merkle": "0xfcf7c365f64b3b366b57f37a40b05932cf23bf8d129accefb97eacea3c7f7424", "prev": "0xc7f1bda8c45e4b284075e3d487cf9168cbbae905dc5a6b2b08da1fce627e766a"}
INFO [08-14|20:04:37.720] Chain reorg detected                     number=262,741 hash=cd2b8e..6dfdc2 drop=1 dropfrom=23a935..770ad2 add=2 addfrom=d207b3..da9204
INFO [08-14|20:04:37.722] Successfully wrote new block with state  number=262,743 hash=d207b3..da9204
INFO [08-14|20:04:37.722] New block in the chain                   "dbft index"=262,743 "chain index"=262,743 hash=0xd207b35ffecc3d369433d342219d7e895ca43dcce8570871c8160824bada9204 "parent hash"=0xc7f1bda8c45e4b284075e3d487cf9168cbbae905dc5a6b2b08da1fce627e766a primary=5 coinbase=0x1212000000000000000000000000000000000003 "mix digest"=0x072bc064323344cba6d63cad4ca88afbea585fc612919e3e351f457ea3704f76
INFO [08-14|20:04:37.722] Fetching latest sealing proposal         "desired number"=262,744
INFO [08-14|20:04:37.724] Commit new sealing work                  number=262,744 sealhash=adf000..a9943a "parent hash"=d207b3..da9204 etherbase=0x1212000000000000000000000000000000000003 txs=0    gas=0 fees=0 elapsed=1.808ms
INFO [08-14|20:04:38.724] Sealing proposal updated                 number=262,744 sealhash=adf000..a9943a "parent hash"=d207b3..da9204 txs=0
2024-08-14T20:04:38.725+0300	INFO	dbft@v0.2.0/dbft.go:107	initializing dbft	{"height": 262744, "view": 0, "index": -1, "role": "WatchOnly"}
2024-08-14T20:04:38.725+0300	DEBUG	dbft@v0.2.0/dbft.go:214	received message	{"type": "Commit", "from": 1, "height": 262743, "view": 0, "my_height": 262744, "my_view": 0}
2024-08-14T20:04:38.725+0300	DEBUG	dbft@v0.2.0/dbft.go:223	ignoring old height	{"height": 262743}
INFO [08-14|20:04:47.567] Looking for peers                        peercount=1 tried=215 static=0
```
Ready for the merge.

**Comment by AnnaShaleva** [comment]
@txhsl, @roman-khimov, please review&merge.

**Review by roman-khimov — APPROVED** [review]
Can't say much about upstream changes, but adaptation is rather clear and if tests are fine then it's OK.

**Review by txhsl — APPROVED** [review]

**Comment by txhsl** [comment]
Will merge this after some necessary tests from our side.

**Pull request merged**